src/node.rs

```diff
@@ -1,37 +1,40 @@
 #[derive(Clone, PartialEq, Debug)]
 pub enum Node {
-    Atom(Atom),
-    List(List),
-    Type(i64)
+    Atom(&'static str),
+    Pair(Box<Node>, Box<Node>),
+    Type(Type),
 }
 
-pub type Atom = &'static str;
-
-// =====================================
-// List
-// =====================================
-pub type List = Vec<Node>;
+#[derive(Clone, PartialEq, Debug)]
+pub enum Type {
+    Universe(i64),
+    Atom,
+    Pair(Box<Type>, Box<Type>)
+}
 
-pub fn cons(list: List, atom: Atom) -> Node {
-    let mut cloned_list = list.clone();
-    cloned_list.push(Node::Atom(atom));
-    Node::List(cloned_list)
+pub fn cons(node: Node, atom: Node) -> Node {
+    Node::Pair(Box::new(node), Box::new(atom))
 }
 
-pub fn car(list: List) -> Node {
-    list.first().unwrap().clone()
+pub fn car(node: Node) -> Result<Node, &'static str> {
+    match node {
+        Node::Pair(a, _) => Result::Ok(*a),
+        _ => Result::Err("Can only apply car to Pair")
+    }
 }
 
-pub fn cdr(list: List) -> Node {
-    let mut cloned_list = list.clone();
-    cloned_list.remove(0);
-    Node::List(cloned_list)
+pub fn cdr(node: Node) -> Result<Node, &'static str> {
+    match node {
+        Node::Pair(_, b) => Result::Ok(*b),
+        _ => Result::Err("Cannot apply cdr to Pair")
```

Comment:
I think it's better to use a dedicated `Error` enum instead of a string for Result::Err... but I digress

Reply:
However, for this case `Option<T>` is better as your return type